pkg/backend/journal.go

```diff
@@ -217,6 +217,19 @@ func (r *JournalReplayer) Add(entry apitype.JournalEntry) {
 		}
 
 		r.base.SecretsProviders = entry.SecretsProvider
+	case apitype.JournalEntryKindRebuiltBaseState:
+		// We need to build the snapshot from the current state here and discard the
+		// current journal entries. This happens after a refresh operation.
+		deployment, _, _ := r.GenerateDeployment()
+		r.base = deployment
```

Comment:
One concern--is it safe to replay the entire journal here to rebuild the base states? It seems like that would only be safe to do if either (a) the journal only consists of mutations of base states or (b) the rebuild is the last entry in the journal. Otherwise, won't the new base states include newly-registered resources, which would throw off the indexing? Feels like this would otherwise need additional coordination from the journal frontend, which I don't see here: https://github.com/pulumi/pulumi/pull/20596/files#diff-33927ecfcea027e3551bdb273868fc780581bc6e92fa3855f6dd3414c13d1842R265-R267

Reply:
(or we need whatever `Rebuild` does to produce the same set of base states as is observed by the generator, and we need the generator to reset its internal state as well)

Reply:
In general, when we're thinking about the journal we need to think about consistency of internal state between the replayer and the generator and the constraints that places on the semantics of each individual entry kind. Then we can define notions of well-formedness for the journal that can be validated, etc. This will help us ensure that we're not generating incorrect results.

Reply:
Added this now.  It's also based on top of #20611 (hopefully to be merged soon, once GitHub actions stops having issues) now, so we can re-use that error reporting infrastructure.